File: app/src/main/java/com/allrecipes/presenters/VideoDetailsScreenPresenter.java
package com.allrecipes.presenters;

import com.allrecipes.managers.GoogleYoutubeApiManager;
import com.allrecipes.model.SearchChannelVideosResponse;
import com.allrecipes.model.YoutubeItem;
import com.allrecipes.model.video.YoutubeVideoResponse;
import com.allrecipes.ui.home.views.HomeScreenView;
import com.allrecipes.ui.videodetails.views.VideoDetailsView;

import java.lang.ref.WeakReference;
import java.util.List;

import io.reactivex.annotations.NonNull;
import io.reactivex.disposables.Disposable;
import io.reactivex.functions.Consumer;

public class VideoDetailsScreenPresenter extends AbstractPresenter<VideoDetailsView> {

    private final GoogleYoutubeApiManager googleYoutubeApiManager;

    Disposable fetchVideoDisposable;

    public VideoDetailsScreenPresenter(
        VideoDetailsView view,
        GoogleYoutubeApiManager googleYoutubeApiManager
    ) {
        super(new WeakReference<>(view));
        this.googleYoutubeApiManager = googleYoutubeApiManager;
    }

    @Override
    public void unbindAll() {
        dispose(fetchVideoDisposable);
    }

    public void fetchVideo(String videoId) {
        getView().showLoading();
        fetchVideoDisposable = googleYoutubeApiManager.fetchVideo(videoId)
            .subscribe(new Consumer<YoutubeVideoResponse>() {
                @Override
                public void accept(@NonNull YoutubeVideoResponse response) throws Exception {
                    if (isDisposedAndViewAvailable(fetchVideoDisposable)) {
                        getView().hideLoading();
                        getView().setVideoDetails(response.items.get(0));
                    }
                }
            }, new Consumer<Throwable>() {
                @Override
                public void accept(@NonNull Throwable throwable) throws Exception {
                    if (isDisposedAndViewAvailable(fetchVideoDisposable)) {
                        getView().hideLoading();
                        throwable.printStackTrace();
                    }
                }
            });

    }
}

File: app/src/main/java/com/allrecipes/presenters/VideoDetailsScreenPresenter.kt
package com.allrecipes.presenters

import com.allrecipes.managers.FavoritesManager
import com.allrecipes.managers.GoogleYoutubeApiManager
import com.allrecipes.managers.remoteconfig.RemoteConfigManager
import com.allrecipes.model.SearchChannelVideosResponse
import com.allrecipes.model.YoutubeItem
import com.allrecipes.model.video.YoutubeVideoResponse
import com.allrecipes.ui.home.views.HomeScreenView
import com.allrecipes.ui.videodetails.views.VideoDetailsView

import java.lang.ref.WeakReference

import io.reactivex.annotations.NonNull
import io.reactivex.disposables.Disposable
import io.reactivex.functions.Consumer

class VideoDetailsScreenPresenter(
    view: WeakReference<VideoDetailsView>,
    private val googleYoutubeApiManager: GoogleYoutubeApiManager,
    private val remoteConfigManager: RemoteConfigManager,
    private val favoritesManager: FavoritesManager
) : AbstractPresenter<VideoDetailsView>(view) {

    private var fetchVideoDisposable: Disposable? = null

    override fun unbindAll() {
        dispose(fetchVideoDisposable)
    }

    fun fetchVideo(video: YoutubeItem) {
        getView().showLoading()
        val id: String
        if (video.id != null && video.id.videoId != null)
            id = video.id.videoId
        else
            id = video.snippet.resourceId.videoId
        fetchVideoDisposable = googleYoutubeApiManager.fetchVideo(id)
            .subscribe({ response ->
               if (isDisposedAndViewAvailable(fetchVideoDisposable)) {
                   getView().hideLoading()
                   getView().setVideoDetails(response.items[0], isSavedAsFavorite(video))
               }
            }) { throwable ->
                if (isDisposedAndViewAvailable(fetchVideoDisposable)) {
                    getView().hideLoading()
                    throwable.printStackTrace()
                    getView().handleApiError(throwable, { fetchVideo(video) })
                }
            }
    }

    fun playVideo() {
        if (remoteConfigManager.isOpenYoutubeNativePlayer) {
            getView().playVideoWithYoutubeNativeAppPlayer()
        } else {
            getView().playVideoWithYoutubeInAppPlayer()
        }
    }

    fun isSavedAsFavorite(videoItem: YoutubeItem): Boolean {
        return favoritesManager.findItemInFavorites(videoItem) != null
    }

    fun removeFavoriteItem(videoItem: YoutubeItem) {
        favoritesManager.removeChannelFavoriteVideo(videoItem)
    }

    fun putFavoriteItem(videoItem: YoutubeItem) {
        favoritesManager.putChannelFavoriteVideo(videoItem)
    }
}
